rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -177,31 +284,52 @@ private static FlowNode getStageExecStartNode(List<FlowNode> allNodesSorted, int
      * @return The name of the node on which the supplied FlowNode executed.
      */
     public static String getExecNodeName(FlowNode flowNode) {
-        String execNodeName = FlowNodeExecutorNameCache.getExecNodeName(flowNode);
+        if (flowNode == null) {
+            return "master";
+        }
+
+        Cache<FlowNode, String> execNodeNameCache = CacheExtension.all().get(0).getExecNodeNameCache();
+        String execNodeName = execNodeNameCache.getIfPresent(flowNode);
         if (execNodeName != null) {
             return execNodeName;
         }
 
-        List<FlowNode> parents = flowNode.getParents();
+        // It is sufficient to walk through just the ancestry
+        ArrayDeque<FlowNode> ancestry = new ArrayDeque<FlowNode>(); // For these nodes, add cache entries
 
-        execNodeName = "master";
-        if (parents != null && !parents.isEmpty()) {
-            // Don't need to iterate all parents. Working back through one ancestral line
-            // should get us to the node containing the WorkspaceAction.
-            FlowNode parent = parents.get(0);
-            WorkspaceAction executorAction = parent.getAction(WorkspaceAction.class);
+        while (flowNode != null) {
+            // Always hit the cache first
+            execNodeName = execNodeNameCache.getIfPresent(flowNode);
+            if (execNodeName != null) {
+                break;
+            }
+
+            // Then we check for the node itself
+            WorkspaceAction executorAction = flowNode.getAction(WorkspaceAction.class);
             if (executorAction != null) {
                 String node = executorAction.getNode();
                 if (node.length() > 0) {
                     execNodeName = node;
                 }
+                break;
+            }
+
+            // Next look at ancestors
+            ancestry.push(flowNode);
+            List<FlowNode> parents = flowNode.getParents();
+            if (parents != null && !parents.isEmpty()) {
+                flowNode = parents.get(0);
             } else {
-                execNodeName = getExecNodeName(parent);
+                break;
             }
         }
-
-        FlowNodeExecutorNameCache.cache(execNodeName, flowNode);
-
+        // Ran on master if no executor workspace
```

Comment:
Please look closely at the original code here - this is the expected handling, you might not agree with it (I can see the logic), but it is not wrong, within the context of this PR (changing that behavior would be out of scope). 

That said, the original implementation (which this is only improving scaling on, until the new FlowAnalyzer is read), was patently broken and did not even attempt to look at block scoping, so... the entirety here is going to be ripped out and we can discuss if the behavior should change then.